### ISSUE TITLE
Reading in a 4M cell coarse grid into parallel::distributed::Triangulation is slow

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#7988]
@ragusa and @loganharbour reported an interesting issue with the `parallel::distributed::Triangulation`.

Namely reading in [1] a large coarse grid with 4M cells takes _significantly longer_ (in wall time and memory consumption) than with the serial `Triangulation`---all happening just on a single core. We are aware of the fact that one should ideally not read in such a large coarse grid. But the increase in computational time is still a bit surprising (we haven't found the time to investigate much further than this).

I will attach a minimal POC (creating a 4M cell triangulation and copying it over to a distributed one) a bit later today.

@tjhei @kronbichler *ping*

[1] This also happens when reading into a serial triangulation first and then copying over to a distributed one.

**Comment by bangerth** [comment]
For a self-reproducing testcase, just start with a mesh with a single cell, refine it a bunch of times, do `GridGenerator::flatten`, and copy the result of this into the parallel triangulation.

**Comment by bangerth** [comment]
@cburstedde -- FYI

When we first wrote the p4est interfaces, we did hit on bottlenecks in p4est that had quadratic behavior in the number of coarse cells. Carsten et al. fixed these, and consequently we now have tests in the testsuite that check that things work sufficiently quickly with 300,000 cells in the coarse mesh. (That's one of the stored external meshes we have in the testsuite.) But we all know that if you increase the size of a problem by a factor of ten, you'll see another bottleneck :-)

**Comment by kronbichler** [comment]
I would say that a factor of 2-3 in increased time is expected because we do certain operations several times plus have p4est do something similar. The interesting question is whether there is actually an increase in complexity when moving to larger grids, which would then be worth investigating.

I think the largest meshes I've worked with are around 1M-3M coarse cells - but these are painful because you cannot really run them in parallel when the mesh on each MPI rank takes 5G or so for the mesh only. That said, together with @peterrum we are working on a concept similar to #3956 (which will be combined with or replace that PR depending on the wishes of @tjhei).

**Comment by cburstedde** [comment]
I agree. If the run time becomes nonlinear in the number of trees, we may want to look at it.

**Comment by tjhei** [comment]
> That said, together with @peterrum we are working on a concept similar to #3956 (which will be combined with or replace that PR depending on the wishes of @tjhei).

I think something of this kind would be great to have. I never got back to working on it, unfortunately. It would make sense to discuss things, if you are interested. I think it would be ideal, if this would not depend on p4est or the p4est space-filling curve. That way we can generate custom partitioning.

**Comment by tamiko** [comment]
Short program demonstrating the problem:

```
#include <deal.II/base/timer.h>

#include <deal.II/distributed/tria.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_in.h>
#include <deal.II/grid/tria.h>

#include <fstream>
#include <iostream>

using namespace dealii;

int
main(int argc, char **argv)
{
  constexpr int dim = 2;
  constexpr unsigned int refinement = 12;

  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);

  dealii::TimerOutput timer(MPI_COMM_WORLD,
                            std::cout,
                            TimerOutput::never,
                            TimerOutput::wall_times);

  Triangulation<dim> triangulation_serial;
  {
    TimerOutput::Scope s(timer, "serial");

    std::cout << "Creating serial triangulation... " << std::flush;
    Triangulation<dim> triangulation_temp;
    GridGenerator::hyper_cube(triangulation_temp);
    triangulation_temp.refine_global(refinement);
    GridGenerator::flatten_triangulation(triangulation_temp,
                                         triangulation_serial);
    std::cout << "done" << std::endl;

#if 0
    std::cout << "Reading serial triangulation... " << std::flush;
    std::string grid_file = "zach_large.ucd";
    GridIn<dim> grid_in;
    grid_in.attach_triangulation(triangulation_serial);
    std::ifstream file(grid_file);
    grid_in.read_ucd(file);
    std::cout << "done." << std::endl;
#endif
  }

  parallel::distributed::Triangulation<dim> triangulation_distributed(
    MPI_COMM_WORLD);
  {
    TimerOutput::Scope s(timer, "parallel");

    std::cout << "Copying to distributed triangulation... " << std::flush;
    triangulation_distributed.copy_triangulation(triangulation_serial);
    std::cout << "done" << std::endl;
  }

  timer.print_summary();

  return 0;
}
```

**Comment by tamiko** [comment]
Some results:
1M (2D, refinement 10):
```
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |      15.4s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| parallel                        |         1 |       7.6s |        49% |
| serial                          |         1 |       7.8s |        51% |
+---------------------------------+-----------+------------+------------+
```
4M (2D, refinement 11):
```
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |       107s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| parallel                        |         1 |        72s |        67% |
| serial                          |         1 |      35.4s |        33% |
+---------------------------------+-----------+------------+------------+
```
16M (2D, refinement 12):
```
(waiting)
```
4M (3D, generated UCD file):
```
(waiting)
```

**Comment by tamiko** [comment]
So we're talking about an algorithmic complexity of about 
O(N^1.08) for serial versus O(N^1.64) in parallel.

**Comment by cburstedde** [comment]
> So we're talking about an algorithmic complexity of about 
> O(N^1.08) for serial versus O(N^1.64) in parallel.

Can you say which p4est functions are used apart from p4est_connectivity_load?

The only functions in p4est_connectivity that use a communicator are _complete
and _bcast.  The others should have a runtime independent of parallel/serial
execution of the surrounding code.

**Comment by tamiko** [comment]
@cburstedde Parallel refers to our parallel distribution with p4est, serial to our serial one (without p4est). The timings are all for a run on one processor. I will try to profile the 4M testcase over night.
